### ISSUE TITLE
[consistent-return] extend base rule with knowledge of return type annotations

### ISSUE THREAD

**Issue by arcanis** [typescript-eslint/typescript-eslint#1987]
**Repro**

```JSON
{
  "rules": {
    "consistent-return": ["error"]
  }
}
```

```TS
function foo(): void {
  // ...
}

function bar(): void {
  if (something)
    return foo();

  // ...
}
```

**Expected Result**

Shouldn't report an error.

**Actual Result**

Reports an error. ESLint doesn't know that `foo` is a function that doesn't have return value, so it enforces that `bar` must end with an explicit return:

```ts
function bar(): void {
  if (something)
    return foo();

  // ...
  return undefined;
}
```

**Versions**

n/a

**Comment by arcanis** [comment]
Indeed! Sorry, both were in a recent PR, I mixed them both 😅

**Comment by Josh-Cena** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/1277
